bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthClientServiceImpl.java

```diff
@@ -0,0 +1,381 @@
+                this.requestParams.isUseBasicAuthentication());
+
+        oauthToken.setCreated(new Date()); // mark created timestamp
+
+        // Put in Store
+        if (!oauthToken.isExpired(new Date(), tokenExpiresInBuffer)) {
+            storageHandler.setOAuthToken(oauthProviderParams, requestParams, oauthToken);
+        }
+
+        return oauthToken;
+    }
+    
+    public void importOauthToken(@NonNull OAuthProviderParams oauthProviderParams, OAuthRequestParams requestParams, OAuthToken oAuthToken) {
+    	storageHandler.setOAuthToken(oauthProviderParams, requestParams, oAuthToken);
+    }
+
+    @Override
+    public void remove(@NonNull OAuthProviderParams oauthProviderParams) {
+        logger.debug("remove");
+        storageHandler.removeAll(oauthProviderParams);
+    }
+
+    private void dependenciesCheck() {
+        logger.debug("storageService {}", storageService);
+        logger.debug("cipherConfigurator {}", cipherConfigurator);
+        logger.debug("trustManager {}", Arrays.deepToString(trustManager));
+    }
+
+    /**
+     * If a storage service is available, it will be used to store passwords,
+     * access tokens, refresh tokens, etc.
+     *
+     * The OAUTH specification calls for confidential storage for access tokens, refresh tokens.
+     * https://tools.ietf.org/html/rfc6749#section-10.4
+     *
+     * @param storageService
+     */
+    @Reference(name = STORAGE_SERVICE, cardinality = ReferenceCardinality.MANDATORY)
+    public void setStorageService(@NonNull StorageService storageService) {
+        this.storageService = storageService;
+        this.storageHandler = new StorageHandler(storageService, OAuthClientServiceImpl.class.getSimpleName(),
+                this.storageEncryptionCipher, this.storageDecryptionCipher);
+    }
+
+    public void unsetStorageService(StorageService storageService) {
+        this.storageService = null;
+    }
+
+    /**
+     * In the authorization grant and access token calls, HTTPS is used.
+     * The TrustManager provides a way to check against the authorization server.
+     * For example, certificate pinning for the authorization server.
+     *
+     * It is used to initialize the SSLContext in HTTPS connections.
+     * 
+     * {@see https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html#init-javax.net.ssl.KeyManager:A-javax.net.ssl.TrustManager:A-java.security.SecureRandom-}
+     *
+     */
+    @Reference(name = OAUTH_X509_TRUST_MANAGER, cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC)
+    public void setTrustManager(@Nullable TrustManager[] trustManager) {
+        this.trustManager = trustManager;
+    }
+
+    public void unsetTrustManager(@Nullable TrustManager[] trustManager) {
+        this.trustManager = null;
+    }
+
+    @Reference(cardinality = ReferenceCardinality.AT_LEAST_ONE, policy = ReferencePolicy.STATIC)
+    public void setCipherConfigurator(CipherConfigurator cipherConfigurator) {
```

Comment:
As far as I know, for 1..n and 0..n cardinality, the naming convention of bind and unbind methods is bindX and unbindX